File: isegoria/src/main/java/com/eulersbridge/isegoria/profile/TaskAdapter.java
package com.eulersbridge.isegoria.profile;

import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.RequestManager;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.network.api.API;
import com.eulersbridge.isegoria.network.api.models.Task;
import com.eulersbridge.isegoria.network.api.responses.PhotosResponse;
import com.eulersbridge.isegoria.util.network.SimpleCallback;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import retrofit2.Response;

class TaskAdapter extends RecyclerView.Adapter<TaskViewHolder> {
    final private RequestManager glide;
    final private API api;
    final private List<Task> items = new ArrayList<>();

    TaskAdapter(@NonNull RequestManager glide, @NonNull API api) {
        this.glide = glide;
        this.api = api;
    }

    void setItems(@NonNull List<Task> newItems) {
        items.clear();
        items.addAll(newItems);
        notifyItemRangeChanged(0, newItems.size());
    }

    @Override
    public int getItemCount() { return items.size(); }

    private int getImageIndex() {
        switch (Resources.getSystem().getDisplayMetrics().densityDpi) {
            case DisplayMetrics.DENSITY_LOW:
                return 5;
            case DisplayMetrics.DENSITY_MEDIUM:
                return 4;
            default:
                return 3;
        }
    }

    @Override
    public void onBindViewHolder(TaskViewHolder viewHolder, int index) {
        Task item = items.get(index);

        viewHolder.setItem(item);

        final int imageIndex = getImageIndex();
        final long itemId = item.id;

        WeakReference<TaskViewHolder> weakViewHolder = new WeakReference<>(viewHolder);

        api.getPhotos(itemId).enqueue(new SimpleCallback<PhotosResponse>() {
            @Override
            protected void handleResponse(Response<PhotosResponse> response) {
                PhotosResponse body = response.body();

                if (body != null
                        && body.totalPhotos > (imageIndex + 1)) {
                    TaskViewHolder innerViewHolder = weakViewHolder.get();

                    if (innerViewHolder != null) {
                        String imageUrl = body.photos.get(imageIndex).thumbnailUrl;

                        if (!TextUtils.isEmpty(imageUrl))
                            innerViewHolder.setImageUrl(glide, itemId, imageUrl);
                    }
                }
            }
        });
    }

    @Override
    public TaskViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        View itemView = LayoutInflater.from(viewGroup.getContext()).
                inflate(R.layout.profile_tasks_list_item, viewGroup, false);
        return new TaskViewHolder(itemView);
    }

    @Override
    public void onViewRecycled(TaskViewHolder holder) {
        holder.onRecycled();

        super.onViewRecycled(holder);
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/profile/TaskAdapter.kt
package com.eulersbridge.isegoria.profile

import android.content.res.Resources
import android.support.v7.widget.RecyclerView
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.ViewGroup
import com.bumptech.glide.RequestManager
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.network.api.API
import com.eulersbridge.isegoria.network.api.models.Task
import com.eulersbridge.isegoria.onSuccess
import java.lang.ref.WeakReference
import java.util.*

internal class TaskAdapter(
    private val glide: RequestManager,
    private val api: API
) : RecyclerView.Adapter<TaskViewHolder>() {
    private val items = ArrayList<Task>()

    fun setItems(newItems: List<Task>) {
        items.clear()
        items.addAll(newItems)
        notifyItemRangeChanged(0, newItems.size)
    }

    override fun getItemCount() = items.size

    private fun getImageIndex(): Int {
        return when (Resources.getSystem().displayMetrics.densityDpi) {
            DisplayMetrics.DENSITY_LOW -> 5
            DisplayMetrics.DENSITY_MEDIUM -> 4
            else -> 3
        }
    }

    override fun onBindViewHolder(viewHolder: TaskViewHolder, index: Int) {
        val item = items[index]

        viewHolder.setItem(item)

        val imageIndex = getImageIndex()
        val itemId = item.id

        val weakViewHolder = WeakReference(viewHolder)
        api.getPhotos(itemId).onSuccess {
            if (it.totalPhotos > imageIndex + 1) {
                val innerViewHolder = weakViewHolder.get()

                if (innerViewHolder != null) {
                    val imageUrl = it.photos?.get(imageIndex)?.thumbnailUrl

                    if (!imageUrl.isNullOrBlank())
                        innerViewHolder.setImageUrl(glide, item.id, imageUrl!!)
                }
            }
        }
    }

    override fun onViewRecycled(holder: TaskViewHolder) {
        holder.onRecycled()

        super.onViewRecycled(holder)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): TaskViewHolder {
        val itemView = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.profile_tasks_list_item, viewGroup, false)
        return TaskViewHolder(itemView)
    }
}